app/services/group_finalizer.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+#
+# Service object to finalize groups
+class GroupFinalizer < Updater
+  def self.finalize(**params)
+    new(**params).finalize
+  end
+
+  # Initialize a GroupFinalizer
+  #
+  # @param group [Group] The group object to be updated
+  def initialize(group:)
+    @group = group
+    @errors = []
+  end
+
+  def finalize
+    return error unless valid?
+    ActiveRecord::Base.transaction do
+      group.finalizing!
+      group.leader.membership.update(locked: true)
+    end
+    success
+  rescue ActiveRecord::RecordInvalid => failures
+    @errors = failures
+    error
+  end
+
+  private
+
+  attr_reader :group, :errors
+
+  def valid?
+    group_full
```

Comment:
Hmm it occurs to me that if you ran `valid?` multiple times that it would just keep adding stuff to the errors array. Not sure if there's a way around this (and I did something similar in `DrawActivator`, but that doesn't mean it's good. Definitely fine for now, but we should open an issue to fix this (maybe call it in the initializet?).